coderd/database/modelmethods.go

```diff
@@ -173,7 +174,44 @@ func (v TemplateVersion) RBACObjectNoTemplate() rbac.Object {
 
 func (g Group) RBACObject() rbac.Object {
 	return rbac.ResourceGroup.WithID(g.ID).
-		InOrg(g.OrganizationID)
+		InOrg(g.OrganizationID).
+		// Group members can read the group.
+		WithGroupACL(map[string][]policy.Action{
+			g.ID.String(): {
+				policy.ActionRead,
+			},
+		})
+}
+
+type GroupMemberRBACHelper struct {
+	User           User
+	GroupID        uuid.UUID
+	OrganizationID uuid.UUID
+}
+
+func (gm GroupMemberRBACHelper) RBACObject() rbac.Object {
+	return rbac.ResourceGroup.WithID(gm.GroupID).InOrg(gm.OrganizationID).
+		// Group member can see they are in the group.
+		WithACLUserList(map[string][]policy.Action{
+			gm.User.ID.String(): {
+				policy.ActionRead,
+			},
+		})
+}
```

Comment:
This is a nifty trick. We can accomplish this same thing without using an ACL though if we just make a new resource for group members. 

What we can do is add a new `group_member` resource to [`policy.go`](https://github.com/coder/coder/blob/main/coderd/rbac/policy/policy.go#L144-L151). 

Then we can use `rbac.ResourceGroupMember.WithID(r.GroupID).InOrg(r.OrganizationID).WithOwner(r.UserID)` that will grant an allow because the `WithOwner()` is essentially the same thing as the User ACL trick you did.

There is going to be some ripple effects of the new resource, such as adding the new resource permission to [`userAdminRole`](https://github.com/coder/coder/blob/main/coderd/rbac/roles.go#L342-L342), [`orgUserAdmin`](https://github.com/coder/coder/blob/main/coderd/rbac/roles.go#L448-L468), and [`orgAdmin`](https://github.com/coder/coder/blob/main/coderd/rbac/roles.go#L387-L405).

----

I think you still need to do the `fetch` business here: https://github.com/coder/coder/blob/main/coderd/database/dbauthz/dbauthz.go#L1393-L1407 because `organization_id` is not present in the returned `database.User`.

Reply:
### Regarding WithOwner

> WithOwner() is essentially the same thing as the User ACL trick you did

I'm having a bit of trouble getting this approach to work. Am I missing something?

The following gives `true` when you run `opa eval --format=pretty -i input.json -d coderd/rbac/policy.rego "data.authz.allow"`

```json
{
    "action": "read",
    "subject": {
        "id": "the_user_id",
        "groups": [],
        "roles": [],
        "scope": {
            "allow_list": [
                "*"
            ],
            "org": {},
            "site": [
                {
                    "action": "*",
                    "negate": false,
                    "resource_type": "*"
                }
            ],
            "user": []
        }
    },
    "object": {
        "acl_group_list": {},
        "acl_user_list": {
            "the_user_id": ["read"]
        },
        "any_org": false,
        "id": "the_org_id",
        "org_owner": "the_org_id",
        "owner": "",
        "type": "group"
    }
}
```

However, the following gives `undefined`:

```json
{
    "action": "read",
    "subject": {
        "id": "the_user_id",
        "groups": [],
        "roles": [],
        "scope": {
            "allow_list": [
                "*"
            ],
            "org": {},
            "site": [
                {
                    "action": "*",
                    "negate": false,
                    "resource_type": "*"
                }
            ],
            "user": []
        }
    },
    "object": {
        "acl_group_list": {},
        "acl_user_list": {},
        "any_org": false,
        "id": "the_org_id",
        "org_owner": "the_org_id",
        "owner": "the_user_id",
        "type": "group"
    }
}
```

### Regarding the `fetch` business

It seems you were looking at the `main` branch judging by the link. [My changes](https://github.com/coder/coder/blob/91ced7b77147d75c2fd6c39e19980c82b321e9a5/coderd/database/dbauthz/dbauthz.go#L1385) include the `fetch` logic. Or did you mean something else?

Reply:
> However, the following gives undefined:

Yes, that will fail (undefined) for 2 reasons. First, you are missing the `member` role which is implied to all Coder users.

https://github.com/coder/coder/blob/main/coderd/rbac/roles.go#L272-L296

It is always appended in the SQL here: https://github.com/coder/coder/blob/main/coderd/database/queries/users.sql#L236-L236

And it grants `*` to all resources in the user's level.(except a subset). So you get effectively this role for the groups resource. 

```
      {
        "name":"member",
        "display_name":"Member",
        "site":[],
        "org":{},
        "user":[
          {
            "negate":false,
            "resource_type":"group",
            "action":"*"
          }
        ]
      }
```

But **this will still fail** because a group is also owned by an organizatnion: `"org_owner": "the_org_id"`. The rego policy requires you to be a [member of an organization](https://github.com/coder/coder/blob/main/coderd/rbac/policy.rego#L168-L175) to do anything on org owned objects. This is mainly so if you remove a member from an org, they can no longer access their own workspace in said org.

So you also need to add the `organization_member` role, which is also [always implied](https://github.com/coder/coder/blob/main/coderd/database/queries/users.sql#L246-L246) if they are in the org.

With these changes, the `input.json` becomes something like: _I removed the unnecessary perms_
```json
{
  "action":"read",
  "subject":{
    "id":"the_user_id",
    "groups":[

    ],
    "roles":[
      {
        "name":"member",
        "display_name":"Member",
        "site":[
        ],
        "org":{
          "the_org_id": []
        },
        "user":[
          {
            "negate":false,
            "resource_type":"*",
            "action":"*"
          }
        ]
      }
    ],
    "scope":{
      "allow_list":[
        "*"
      ],
      "org":{

      },
      "site":[
        {
          "action":"*",
          "negate":false,
          "resource_type":"*"
        }
      ],
      "user":[

      ]
    }
  },
  "object":{
    "acl_group_list":{

    },
    "acl_user_list":{

    },
    "any_org":false,
    "id":"the_org_id",
    "org_owner":"the_org_id",
    "owner":"the_user_id",
    "type":"group"
  }
}
```

-------

On the fetch business, :facepalm:. I checked out your branch, but the GitLink extension does not work with multiple git origins I suppose. I updated the link to your fork.

Reply:
Ah, ok. Thank you for the clarification on RBAC, that's very helpful.

However, I still don't quite understand the comment regarding `fetch` even with the updated link. RBAC filtering that respects org id is done on [this line](https://github.com/hugodutka/coder/blob/91ced7b77147d75c2fd6c39e19980c82b321e9a5/coderd/database/dbauthz/dbauthz.go#L1408). The returned `[]database.User` is already filtered.

Reply:
@hugodutka yes. I just mean't, I think you still have to write a custom `fetch` function that returns a custom type with all the fields `user_id`, `organizaton_id`, and `id`. 

Unless the custom `view` we talked about in the other comment joins in this information, and we return a `GroupMember` instead of a `database.User`

Reply:
That's clear now. Returning `GroupMember`s from the new `view` instead of `User`s is a more elegant approach imo.